autopush-common/src/db/redis/redis_client/mod.rs

```diff
@@ -0,0 +1,855 @@
+    /// to save messages, but there are other considerations (e.g. mutation limits)
+    async fn save_messages(&self, uaid: &Uuid, messages: Vec<Notification>) -> DbResult<()> {
+        // plate simple way of solving this:
+        for message in messages {
+            self.save_message(uaid, message).await?;
+        }
+        Ok(())
+    }
+
+    /// Delete expired messages
+    async fn increment_storage(&self, uaid: &Uuid, timestamp: u64) -> DbResult<()> {
+        let uaid = Uaid(uaid);
+        debug!("🐰🔥 Incrementing storage to {}", timestamp);
+        let msg_list_key = self.message_list_key(&uaid);
+        let exp_list_key = self.message_exp_list_key(&uaid);
+        let mut con = self.connection().await?;
+        let exp_id_list: Vec<String> = con.zrangebyscore(&exp_list_key, 0, timestamp).await?;
+        if exp_id_list.len() > 0 {
+            trace!("🐰🔥 Deleting {} expired msgs", exp_id_list.len());
+            redis::pipe()
+                .del(&exp_id_list)
+                .zrem(&msg_list_key, &exp_id_list)
+                .zrem(&exp_list_key, &exp_id_list)
+                .exec_async(&mut con)
+                .await?;
+        }
+        Ok(())
+    }
+
+    /// Delete the notification from storage.
+    async fn remove_message(&self, uaid: &Uuid, chidmessageid: &str) -> DbResult<()> {
+        let uaid = Uaid(uaid);
+        trace!(
+            "🐰 attemping to delete {:?} :: {:?}",
+            uaid.to_string(),
+            chidmessageid
+        );
+        let msg_key = self.message_key(&uaid, &chidmessageid);
+        let msg_list_key = self.message_list_key(&uaid);
+        let exp_list_key = self.message_exp_list_key(&uaid);
+        debug!("🐰🔥 Deleting message {}", &msg_key);
+        let mut con = self.connection().await?;
+        // We remove the id from the exp list at the end, to be sure
+        // it can't be removed from the list before the message is removed
+        redis::pipe()
+            .del(&msg_key)
+            .zrem(&msg_list_key, &chidmessageid)
+            .zrem(&exp_list_key, &chidmessageid)
+            .exec_async(&mut con)
+            .await?;
+        self.metrics
+            .incr_with_tags("notification.message.deleted")
+            .with_tag("database", &self.name())
+            .send();
+        Ok(())
+    }
+
+    /**
+    Topic messages are handled as other messages with redis, we return nothing.
+    */
+    async fn fetch_topic_messages(
+        &self,
+        _uaid: &Uuid,
+        _limit: usize,
+    ) -> DbResult<FetchMessageResponse> {
+        Ok(FetchMessageResponse {
+            messages: vec![],
+            timestamp: None,
+        })
+    }
+
+    /// Return [`limit`] messages pending for a [`uaid`] that have a record timestamp
+    /// after [`timestamp`] (millisecs).
+    ///
+    /// If [`limit`] = 0, we fetch all messages after [`timestamp`].
+    ///
+    /// This can return expired messages, following bigtables behavior
```

Comment:
Yeah, but there's no real need to do that, and in fact, it's probably an anti-pattern we need to fix in bigtable. (There's a potential abuse vector that should be cut off.)

Ideally, messages should be dropped once they've been decisively handled (e.g. if they're ACK'd or if the NACK response specifies that the message is undeliverable. see https://github.com/mozilla-services/autopush-rs/pull/795 )

Reply:
Expired messages are removed from redis, so it doesn't use memory for expired messages.

But it has to return dummy expired events to avoid `check_storage_advance` to stop [if there are 10 expired messages](https://github.com/mozilla-services/autopush-rs/blob/a6fc8cdb678e4bc538e68bab601f801ef8c4b2af/autoconnect/autoconnect-ws/autoconnect-ws-sm/src/identified/on_server_notif.rs#L109) (so messages is an empty vec)

> Ideally, messages should be dropped once they've been decisively handled (e.g. if they're ACK'd or if the NACK response specifies that the message is undeliverable. see https://github.com/mozilla-services/autopush-rs/pull/795 )

`remove_message` do remove the message, so this is already the case :)